addons/l10n_es_edi_facturae/models/account_edi_format.py

```diff
@@ -0,0 +1,498 @@
+        str_tree = re.sub('\n+', '', str_tree)
+        str_tree = re.sub(r'\s{2, }', '', str_tree)
+        if as_string:
+            return str_tree
+        return etree.XML(str_tree)
+
+    @staticmethod
+    def l10n_es_edi_facturae_inv_lines_to_items(invoice, conv_rate, needs_equivalency, fmt_in_cur, fmt_in_eur):
+        """
+        Convert the invoice lines to a list of items required for the Facturae xml generation
+
+        :param account.move invoice:    The invoice containing the invoice lines
+        :param float conv_rate:         The conversion rate from the invoice currency to the company currency
+        :param bool needs_equivalency:  True if the invoice uses multi-currencies
+        :param func fmt_in_cur:         The formatting function to use for the amount in the invoice currency
+        :param func fmt_in_eur:         The formatting function to use for the amount in euros
+        :return tuple:                  A tuple containing the Face items, the taxes and the invoice totals data.
+        """
+
+        def compute_tax_amount(tax, line):
+            if tax.amount_type == 'percent':
+                return tax.amount * line.price_subtotal / 100
+            if tax.amount_type == 'fixed':
+                return tax.amount
+            if tax.amount_type == 'division':
+                if 1 + (tax.amount / 100):
+                    return line.price_subtotal - (line.price_subtotal / (1 + (tax.amount / 100)))
+                return 0.0
+
+        items = []
+        totals = {
+            'total_gross_amount': 0.,
+            'total_general_discounts': 0.,
+            'total_general_surcharges': 0.,
+            'total_taxes_withheld': 0.,
+            'total_payments_on_account': 0.,
+            'amounts_withheld': 0.,
+        }
+        taxes = []
+        for line in invoice.invoice_line_ids:
+            invoice_line_values = EDI_XML_DICT_TEMPLATES['invoice_line'].copy()
+
+            total_price = line.quantity * line.price_unit * conv_rate  # This isn't a mistake
+            gross_amount = line.price_subtotal / conv_rate
+            discount = max(0., line.discount * total_price)
+            surcharge = max(0., - line.discount * total_price)  # + line.price_total - line.price_subtotal
+            totals['total_gross_amount'] += gross_amount
+            totals['total_general_discounts'] += discount
+            totals['total_general_surcharges'] += surcharge
+
+            taxes_output = [{
+                "TaxTypeCode": tax.l10n_es_edi_facturae_tax_type,
+                "TaxRate": f'{tax.amount:.3f}',
+                "TaxableBase": {
+                    'TotalAmount': fmt_in_cur(gross_amount),
+                    'EquivalentInEuros': fmt_in_eur(line.price_subtotal) if needs_equivalency else False,
+                },
+                "TaxAmount": {
+                    "TotalAmount": fmt_in_cur(compute_tax_amount(tax, line) * conv_rate),
```

Comment:
For this we have the tax line details stuff...  Or even the _compute_all method, but this seems like a reinvention of the tax engine.